[proposal] Run <flutter pub get> and update .packages file once Flutter channel is changed in Preferences

## Steps to Reproduce

1. Make sure your current Flutter channel is Stable in IDE Preferences!
2.  Run `flutter pub get`
3.  See <your-project>/.packages file and verify that is uses the stable channel
![Screen Shot 2021-01-28 at 12 08 31](https://user-images.githubusercontent.com/67424416/106130374-8c63ba00-6161-11eb-8236-85633593740b.png)

1. Now go to Preferences and change the Flutter channel to Master

Expected: for the .package file to use the new Flutter channel directory
Actual: This file remains unchanged and stay in a invalid state since it references a invalid Flutter directory that is not in sync with the IDE. 

Now, if you try to run `flutter run` in the Terminal of the IDE, it will show supported devices from Master channel and if you select one of them (Chrome or any desktop target) it will crash because the tool will try to use the .packages file but since it still referencing Stable channel the tool then crash. See the logs bellow


<details>
<summary>logs</summary>

```bash
MacBook-Pro-de-MacBook:flutter_example pedromassango$ flutterm run
Multiple devices found:
macOS (desktop) • macos  • darwin-x64     • macOS 11.1 20C69 darwin-x64
Chrome (web)    • chrome • web-javascript • Google Chrome 88.0.4324.96
[1]: macOS (macos)
[2]: Chrome (chrome)
Please choose one (To quit, press "q/Q"): 1

Launching lib/main.dart on macOS in debug mode...
Warning: CocoaPods recommended version 1.10.0 or greater not installed.
Pods handling may fail on some projects involving plugins.
To upgrade see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.



Running pod install...                                           2,242ms
../../../Code/flutter_stable/packages/flutter/lib/src/widgets/media_query.dart:129:29: Error: Type 'ui.Window' not found.
  MediaQueryData.fromWindow(ui.Window window)                           
                            ^^^^^^^^^                                   
../../../Code/flutter_stable/packages/flutter/lib/src/foundation/binding.dart:88:3: Error: Type 'ui.Window' not found.
  ui.Window get window => ui.window;                                    
  ^^^^^^^^^                                                             
../../../Code/flutter_stable/packages/flutter/lib/src/rendering/view.dart:60:14: Error: Type 'ui.Window' not found.
    required ui.Window window,                                          
             ^^^^^^^^^                                                  
../../../Code/flutter_stable/packages/flutter/lib/src/rendering/view.dart:88:9: Error: Type 'ui.Window' not found.
  final ui.Window _window;                                              
        ^^^^^^^^^                                                       
../../../Code/flutter_stable/packages/flutter/lib/src/material/page.dart:48:15: Error: Getter not found: 'opaque'.
       assert(opaque),                                                  
              ^^^^^^                                                    
../../../Code/flutter_stable/packages/flutter/lib/src/material/page.dart:189:15: Error: Getter not found: 'opaque'.
       assert(opaque),                                                  
              ^^^^^^                                                    
../../../Code/flutter_stable/packages/flutter/lib/src/widgets/framework.dart:1900:15: Error: Getter not found: 'runtimeType'.
            "$runtimeType's children must not contain any null values, "
              ^^^^^^^^^^^                                               
../../../Code/flutter_stable/packages/flutter/lib/src/widgets/media_query.dart:129:32: Error: 'Window' isn't a type.
  MediaQueryData.fromWindow(ui.Window window)                           
                               ^^^^^^                                   
../../../Code/flutter_stable/packages/flutter/lib/src/cupertino/route.dart:353:15: Error: Getter not found: 'opaque'.
       assert(opaque),                                                  
              ^^^^^^                                                    
../../../Code/flutter_stable/packages/flutter/lib/src/cupertino/route.dart:380:15: Error: Getter not found: 'opaque'.
       assert(opaque),                                                  
              ^^^^^^                                                    
../../../Code/flutter_stable/packages/flutter/lib/src/rendering/view.dart:60:17: Error: 'Window' isn't a type.
    required ui.Window window,                                          
                ^^^^^^                                                  
../../../Code/flutter_stable/packages/flutter/lib/src/rendering/view.dart:88:12: Error: 'Window' isn't a type.
  final ui.Window _window;                                              
           ^^^^^^                                                       
                                                                        
Command PhaseScriptExecution failed with a nonzero exit code            
note: Using new build system                                            
note: Building targets in parallel                                      
note: Planning build                                                    
note: Constructing build description                                    
** BUILD FAILED **                                                      
                                                                        
Building macOS application...                                           
Exception: Build process failed
MacBook-Pro-de-MacBook:flutter_example pedromassango$ 
```
</details>



## Version info

<details>
<summary>flutter doctor -v</summary>

```bash
[✓] Flutter (Channel master, 1.26.0-13.0.pre.229, on macOS 11.1 20C69 darwin-x64, locale en)
    • Flutter version 1.26.0-13.0.pre.229 at /Users/pedromassango/Code/flutter_master
    • Framework revision 650b240634 (72 minutes ago), 2021-01-27 21:39:06 -0800
    • Engine revision 68cd4c42b5
    • Dart version 2.12.0 (build 2.12.0-267.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/pedromassango/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] IntelliJ IDEA Community Edition (version 2020.3.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 52.2.5
    • Dart plugin version 203.6912

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 11.1 20C69 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 88.0.4324.96

! Doctor found issues in 1 category.
```

</details>

--- a/src/io/flutter/sdk/FlutterSettingsConfigurable.java
+++ b/src/io/flutter/sdk/FlutterSettingsConfigurable.java
@@ -35,6 +35,8 @@ import io.flutter.*;
 import io.flutter.bazel.Workspace;
 import io.flutter.bazel.WorkspaceCache;
 import io.flutter.font.FontPreviewProcessor;
+import io.flutter.pub.PubRoot;
+import io.flutter.pub.PubRoots;
 import io.flutter.settings.FlutterSettings;
 import org.jetbrains.annotations.Nls;
 import org.jetbrains.annotations.NotNull;
@@ -46,6 +48,8 @@ import javax.swing.text.JTextComponent;
 import java.awt.datatransfer.StringSelection;
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.util.List;
+import java.util.concurrent.Semaphore;
 
 // Note: when updating the settings here, update FlutterSearchableOptionContributor as well.
 
@@ -85,6 +89,13 @@ public class FlutterSettingsConfigurable implements SearchableConfigurable {
   private boolean ignoringSdkChanges = false;
 
   private String fullVersionString;
+  private FlutterSdkVersion previousSdkVersion;
+
+  /**
+   * Semaphore used to synchronize flutter commands so we don't try to do two at once.
+   */
+  private final Semaphore lock = new Semaphore(1, true);
+  private Process updater;
 
   FlutterSettingsConfigurable(@NotNull Project project) {
     this.myProject = project;
@@ -95,6 +106,10 @@ public class FlutterSettingsConfigurable implements SearchableConfigurable {
   }
 
   private void init() {
+    final FlutterSdk sdk = FlutterSdk.getFlutterSdk(myProject);
+    if (sdk != null) {
+      previousSdkVersion = sdk.getVersion();
+    }
     mySdkCombo.getComboBox().setEditable(true);
 
     myCopyButton.setSize(ActionToolbar.DEFAULT_MINIMUM_BUTTON_SIZE);
@@ -264,13 +279,22 @@ public class FlutterSettingsConfigurable implements SearchableConfigurable {
 
       final String sdkHomePath = getSdkPathText();
       if (FlutterSdkUtil.isFlutterSdkHome(sdkHomePath)) {
+
         ApplicationManager.getApplication().runWriteAction(() -> {
           FlutterSdkUtil.setFlutterSdkPath(myProject, sdkHomePath);
           FlutterSdkUtil.enableDartSdk(myProject);
+
           ApplicationManager.getApplication().executeOnPooledThread(() -> {
             final FlutterSdk sdk = FlutterSdk.forPath(sdkHomePath);
             if (sdk != null) {
-              sdk.queryFlutterChannel(false);
+              try {
+                lock.acquire();
+                sdk.queryFlutterChannel(false);
+                lock.release();
+              }
+              catch (InterruptedException e) {
+                // do nothing
+              }
             }
           });
         });
@@ -320,6 +344,27 @@ public class FlutterSettingsConfigurable implements SearchableConfigurable {
     }
 
     onVersionChanged();
+    if (sdk != null) {
+      if (previousSdkVersion != null) {
+        if (previousSdkVersion.compareTo(sdk.getVersion()) != 0) {
+          final List<PubRoot> roots = PubRoots.forProject(myProject);
+          try {
+            lock.acquire();
+            for (PubRoot root : roots) {
+              sdk.startPubGet(root, myProject);
+            }
+            lock.release();
+          }
+          catch (InterruptedException e) {
+            // do nothing
+          }
+          previousSdkVersion = sdk.getVersion();
+        }
+      }
+    }
+    else {
+      previousSdkVersion = null;
+    }
 
     myReportUsageInformationCheckBox.setSelected(FlutterInitializer.getCanReportAnalytics());
 
@@ -375,17 +420,40 @@ public class FlutterSettingsConfigurable implements SearchableConfigurable {
       return;
     }
 
+    // Moved launching the version updater to a background thread to avoid deadlock
+    // when the semaphone was locked for a long time on the EDT.
     final ModalityState modalityState = ModalityState.current();
+    ApplicationManager.getApplication().executeOnPooledThread(() -> {
+      try {
+        if (updater != null) {
+          // If we get back here before the previous one finished then just kill it.
+          // This isn't perfect, but does help avoid printing this message most times:
+          // Waiting for another flutter command to release the startup lock...
+          updater.destroy();
+          lock.release();
+        }
+        Thread.sleep(100L);
+        lock.acquire();
 
-    // TODO(devoncarew): Switch this to expecting json output.
-    sdk.flutterVersion().start((ProcessOutput output) -> {
-      final String fullVersionText = output.getStdout();
-      fullVersionString = fullVersionText;
+        ApplicationManager.getApplication().invokeLater(() -> {
+          // "flutter --version" can take a long time on a slow network.
+          updater = sdk.flutterVersion().start((ProcessOutput output) -> {
+            fullVersionString = output.getStdout();
+            final String[] lines = StringUtil.splitByLines(fullVersionString);
+            final String singleLineVersion = lines.length > 0 ? lines[0] : "";
 
-      final String[] lines = StringUtil.splitByLines(fullVersionText);
-      final String singleLineVersion = lines.length > 0 ? lines[0] : "";
-      ApplicationManager.getApplication().invokeLater(() -> updateVersionTextIfCurrent(sdk, singleLineVersion), modalityState);
-    }, null);
+            ApplicationManager.getApplication().invokeLater(() -> {
+              updater = null;
+              lock.release();
+              updateVersionTextIfCurrent(sdk, singleLineVersion);
+            }, modalityState);
+          }, null);
+        }, modalityState);
+      }
+      catch (InterruptedException e) {
+        // do nothing
+      }
+    });
   }
 
   /***

